### PR TITLE
Removed BASE_DIR in favor of CONFIG_DIR and DATA_DIR

### LINKED ISSUE #398: Move config folder to $XDG_CONFIG_HOME
To unclutter everyone's $HOME dir, I would propose usage of the system var XDG_CONFIG_HOME (~/.config usually)
Or to be able to start sup with;
`sup --config ~/.config/sup` so `alias sup=sup --config ~/.config/sup` would be available.

Would this break sup for any non-unix users?

### PR DESCRIPTION
**Pull request by ZJvandeWeg**
Proposed changes to close #398 

As suggested, does everything automagicly.

### CODE REVIEW

File: lib/sup.rb [hunk 1]
```diff
@@ -46,20 +46,38 @@ def yaml_properties *props
...
-  LOG_FN     = File.join(BASE_DIR, "log")
-  SYNC_OK_FN = File.join(BASE_DIR, "sync-back-ok")
+  def xdg_config_dir
+    ENV['SUP_BASE'] if ENV['SUP_BASE']
+
+    dir = ENV['XDG_CONFIG_HOME'] || File.join(ENV['HOME'], ".config")
+    File.join(dir, "sup")
+  end
+
+  def xdg_data_dir
```
Comment (gauteh) [inline lib/sup.rb]: These functions should not be duplicated.

Comment (ZJvandeWeg) [inline lib/sup.rb]: @gauteh 
Do you mean that there should just be one function returning the both of them? 
Or just one conditional statement setting the CONFIG_DIR and DATA_DIR? 
Or the duplication in e.g. /bin/sup?

File: bin/sup-dump [hunk 2]
```diff
@@ -7,7 +7,12 @@ require 'xapian'
 require 'trollop'
 require 'set'
 
-BASE_DIR = ENV["SUP_BASE"] || File.join(ENV["HOME"], ".sup")
+if ENV["SUP_BASE"]
+  DATA_DIR    = ENV['SUP_BASE']
+else
+  dir = ENV['XDG_DATA_HOME'] || File.join(ENV['HOME'], ".local/share")
+  DATA_DIR = File.join(dir, sup)
```
Comment (rakoo) [inline bin/sup-dump]: I believe that should be

```ruby
DATA_DIR = File.join(dir, 'sup')
```

(ie string, not variable)

Comment (ZJvandeWeg) [inline bin/sup-dump]: Ouch. You're absolutely right.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: lib/sup.rb [hunk 3]
```diff
@@ -46,20 +46,36 @@ def yaml_properties *props
...
-  HOOK_DIR   = File.join(BASE_DIR, "hooks")
-  SEARCH_FN  = File.join(BASE_DIR, "searches.txt")
-  LOG_FN     = File.join(BASE_DIR, "log")
-  SYNC_OK_FN = File.join(BASE_DIR, "sync-back-ok")
+  if ENV['SUP_BASE']
+    CONFIG_DIR = ENV['SUP_BASE']
+    DATA_DIR = ENV['SUP_BASE']
+  else
+   CONFIG_DIR = if ENV['XDG_CONFIG_HOME'] 
+		  File.join(ENV['XDG_CONFIG_DIR'], "sup")
```
Comment (rakoo) [inline lib/sup.rb]: Typo: this should be `ENV['XDG_CONFIG_HOME']`

File: lib/sup.rb [hunk 4]
```diff
@@ -46,20 +46,36 @@ def yaml_properties *props
...
+		else
+		  File.join(ENV['HOME'], ".config/sup")
+		end
+   DATA_DIR = 	if ENV['XDG_DATA_HOME']
+		  File.join(['XDG_DATA_HOME'], "sup")
```
Comment (rakoo) [inline lib/sup.rb]: Typo: this should be ENV['XDG_DATA_HOME']

File: bin/sup [hunk 5]
```diff
@@ -18,7 +18,40 @@ end
...
+
+if Dir.exist?((old_sup_dir = File.join(ENV["HOME"], ".sup"))) && !ENV['SUP_BASE']
+  FileUtils.mkdir_p xdg_config_dir
+  FileUtils.mkdir_p xdg_data_dir
+
+  FileUtils.cd old_sup_dir do
+    files = Dir.entries old_sup_dir
+    %w(config.yaml sources.yaml colors.yaml).each do |file|
+      FileUtils.mv file, xdg_config_dir if files.include? file
+    end
```
Comment (rakoo) [inline bin/sup]: My custom coloschemes weren't moved with this, I suggest adding something like that:

```ruby
Dir.glob("colors.*.yaml") do |file|
  FileUtils.mv file, xdg_config_dir
end
```

**Comment by rakoo** [comment]
:+1:

**Comment by ZJvandeWeg** [comment]
@rakoo 
Thanks for the review!

Now your custom color files are moved also. Could you confirm?

**Comment by rakoo** [comment]
Yep, it did catch my colorschemes (nice catch with the `.bak` !)

However it doesn't catch the `sync-back-ok` from the sync-back functionality (https://github.com/sup-heliotrope/sup/wiki/Using-sup-with-other-clients). I believe this one should also go in config.

I'm currently editing the wiki for all the paths involved, to be pushed when we have a new release.

**Comment by ZJvandeWeg** [comment]
Isn't it more off a status file? As `config.yaml` has both the `sync_back_to_maildir` and `sync_back` fields? And the file only exists for data persistance right? 
Im happy to add it if Im really wrong here.

**Comment by ZJvandeWeg** [comment]
Is there something stopping this PR from being merged?

**Comment by gauteh** [comment]
Maybe there should be a check if whether the .sup dir is empty now? In case users have some kind of special maildir-setup or hook setup that involves .sup? then let sup fail to start if the dir is not empty after the move, asking the user to complete the move, or ignore it with a config option?

**Comment by ZJvandeWeg** [comment]
The `.sup` dir is empty. 
First all known sup conf be a check if whether the .sup dir is empty now? In case users have some kind of special maildir-setup or hook setup that involves .supig files are moved to `XDG_CONFIG`. Then all other contents of the folder is moved to `XDG_DATA`. See line 51 of `bin/sup`.

Or am I not really getting your question?

**Comment by gauteh** [comment]
Ah, ok. I didn't realize you moved all the rest to DATA. I guess it
should be ok, just a little bit anxious about this PR since it has
potential for .. mishaps.

One weird case I though of is if a user has stored his mail in .sup, so
that sources and fetching scripts will have to be changed. With
offlineimap the result could be deleted mail.



ZJvandeWeg writes on July 31, 2015 9:47:

**Comment by gauteh** [comment]
Perhaps we could ask the user on start-up if he wants to have the data
moved? And also stop once more after moving so that manual changes could
be done before polling is started.

ZJvandeWeg writes on July 31, 2015 9:47:

**Comment by gauteh** [comment]
What I was thinking about is something like this: https://github.com/gauteh/sup/tree/xdg-dev ? If you agree, please merge it into your branch.

**Comment by ZJvandeWeg** [comment]
Looks good, will merge today and update this PR.

**Comment by ZJvandeWeg** [comment]
@gauteh Merged!

**Comment by gauteh** [comment]
Thanks! Squashed and merged!

**Pull request closed without merging**